File: examples/bind.go
package main

import (
	"time"

	"github.com/webview/webview"
)

const html = `<button id="increment">Tap me</button>
<div>You tapped <span id="count">0</span> time(s).</div>
<button id="compute">Compute</button>
<div>Result of computation: <span id="compute-result">0</span></div>
<script>
  const [incrementElement, countElement, computeElement, computeResultElement] =
    document.querySelectorAll("#increment, #count, #compute, #compute-result");
  document.addEventListener("DOMContentLoaded", () => {
    incrementElement.addEventListener("click", () => {
      window.increment().then(result => {
        countElement.textContent = result.count;
      });
    });
    computeElement.addEventListener("click", () => {
      computeElement.disabled = true;
      window.compute(6, 7).then(result => {
		console.log(result);
        computeResultElement.textContent = result;
        computeElement.disabled = false;
      });
    });
  });
</script>`

type IncrementResult struct {
	Count uint `json:"count"`
}

func main() {
	var count uint = 0
	w := webview.New(false)
	defer w.Destroy()
	w.SetTitle("Bind Example")
	w.SetSize(480, 320, webview.HintNone)

	// A binding that increments a value and immediately returns the new value.
	w.Bind("increment", func() IncrementResult {
		count++
		return IncrementResult{Count: count}
	})

	// A binding that performs a long-running computation and returns the result
	w.Bind("compute", func(a, b int) chan webview.BindCallbackResult {
		ch := make(chan webview.BindCallbackResult)
		go func() {
			defer close(ch)
			time.Sleep(1 * time.Second)
			ch <- webview.BindCallbackResult{Value: a * b, Error: nil}
		}()
		return ch
	})

	w.SetHtml(html)
	w.Run()
}

File: examples/bind.cc
#include "bind_lib.h"
#include "webview.h"
#include <chrono>
#include <iostream>

#ifdef _WIN32
int WINAPI WinMain(HINSTANCE /*hInst*/, HINSTANCE /*hPrevInst*/,
                   LPSTR /*lpCmdLine*/, int /*nCmdShow*/) {
#else
int main() {
#endif
  try {
    webview_cc wv(true, nullptr);
    int count = 0;

    wv.set_title("Bind Example");
    wv.set_size(480, 320, WEBVIEW_HINT_NONE);

    // A binding that counts up or down and immediately returns the new value.
    wv.bind(
        "count",
        [&](const std::string &id, const std::string &req, void * /*arg*/) {
          // We use the inbuilt Webview JSON utility from webview::strings::json
          auto increment = std::stoi(json::parse(req, "", 0));
          wv.resolve(id, 0, count += increment);
        },
        nullptr);

    // A binding that does computational work and returns the result at a later time.
    wv.bind(
        "compute",
        [&](const std::string &id, const std::string & /*req*/,
            void * /*arg*/) {
          // Simulate load.
          std::this_thread::sleep_for(std::chrono::seconds(3));
          wv.resolve(id, 0, answer_machine_result());
        },
        nullptr);

    wv.set_html(WEBVIEW_BIND_EXAMPLE_HTML);
    wv.run();
  } catch (const webview::errors::exception &e) {
    std::cerr << e.what() << '\n';
    return 1;
  }

  return 0;
}
